### PR TITLE
feature: support ConsistentHashLoadBalance

### PR DESCRIPTION
**Pull request by ph3636**
### Ⅰ. Describe what this PR did
support ConsistentHashLoadBalance

### CODE REVIEW

File: config/seata-config-core/src/main/java/io/seata/config/ConfigurationKeys.java [hunk 1]
```diff
@@ -41,4 +41,12 @@
      * The constant FILE_ROOT_TYPE.
      */
     public static final String FILE_ROOT_TYPE = "type";
+    /**
+     * The constant LOAD_BALANCE.
+     */
+    public static final String LOAD_BALANCE = FILE_ROOT_REGISTRY + FILE_CONFIG_SPLIT_CHAR + "load-balance";
```
Comment (zjinlei) [inline config/seata-config-core/src/main/java/io/seata/config/ConfigurationKeys.java]: please use camel style all

Comment (ph3636) [reply config/seata-config-core/src/main/java/io/seata/config/ConfigurationKeys.java]: done

File: discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java [hunk 2]
```diff
@@ -24,12 +28,23 @@
...
+            ConfigurationFactory.CURRENT_FILE_INSTANCE.getConfig(LOAD_BALANCE));
+
     /**
      * Gets instance.
      *
      * @return the instance
      */
     public static LoadBalance getInstance() {
-        return EnhancedServiceLoader.load(LoadBalance.class);
+        return LOAD_BALANCE_INSTANCE;
```
Comment (l81893521) [inline discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java:48]: I think can return
```
EnhancedServiceLoader.load(LoadBalance.class,
            ConfigurationFactory.CURRENT_FILE_INSTANCE.getConfig(LOAD_BALANCE));
```
directly, cause EnhancedServiceLoader already support singleton.

File: discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java [hunk 3]
```diff
@@ -24,12 +28,23 @@
...
 
+    /**
+     * The constant LOAD_BALANCE.
+     */
+    private static final String LOAD_BALANCE = FILE_ROOT_REGISTRY + FILE_CONFIG_SPLIT_CHAR + "loadBalance";
+
+    /**
+     * The load balance.
+     */
+    private static final LoadBalance LOAD_BALANCE_INSTANCE = EnhancedServiceLoader.load(LoadBalance.class,
```
Comment (slievrly) [inline discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java:39]: If the configuration file does not have this configuration item, a default value is required. Compatible with previous versions

File: discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java (hunk 2 above)
Comment (ph3636) [reply discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java]: done

File: discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java (hunk 3 above)
Comment (ph3636) [reply discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java]: done

File: discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java [hunk 4]
```diff
@@ -25,11 +29,21 @@
...
     /**
-     * Gets instance.
+     * The constant LOAD_BALANCE.
+     */
+    private static final String LOAD_BALANCE = FILE_ROOT_REGISTRY + FILE_CONFIG_SPLIT_CHAR + "loadBalance";
+
+    /**
+     * The default load balance.
+     */
+    private static final String DEFAULT_LOAD_BALANCE = "RandomLoadBalance";
```
Comment (slievrly) [inline discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java:39]: remove to DefaultValues?

Comment (ph3636) [reply discovery/seata-discovery-core/src/main/java/io/seata/discovery/loadbalance/LoadBalanceFactory.java]: done

**Comment by slievrly** [comment]
@ph3636 There are cyclic dependencies.

**Comment by ph3636** [comment]
> @ph3636 There are cyclic dependencies.

done

**Comment by ls9527** [comment]
根据TC列表做一层缓存吧， 每次都新建虚拟节点是不是性能会受到影响？ 比如配置3台TC， 配置160个虚拟节点，  每个请求都要多480个循环， 听起来就没必要嘛

**Pull request merged**